presto-main/src/main/java/io/prestosql/spiller/AesSpillCipher.java

```diff
@@ -0,0 +1,193 @@
+    private final int ivBytes;
+
+    AesSpillCipher()
+    {
+        this.key = generateNewSecretKey();
+        this.ivBytes = createEncryptCipher().getIV().length;
+    }
+
+    @Override
+    public ByteBuffer decrypt(ByteBuffer encryptedData)
+    {
+        int length = encryptedData.remaining();
+        if (length < ivBytes) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Encrypted data size is too small: %s. It must be at least the size of the initialization vector: %s.", length, ivBytes));
+        }
+        byte[] iv = new byte[ivBytes];
+        encryptedData.get(iv);
+        Cipher cipher = createDecryptCipher(new IvParameterSpec(iv));
+        ByteBuffer output = ByteBuffer.allocate(cipher.getOutputSize(encryptedData.remaining()));
+        try {
+            cipher.doFinal(encryptedData, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Cannot decrypt previously encrypted data: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public byte[] decrypt(byte[] encryptedData)
+    {
+        return unwrapOrCopyToByteArray(decrypt(ByteBuffer.wrap(encryptedData)));
+    }
+
+    @Override
+    public ByteBuffer encrypt(ByteBuffer data)
+    {
+        Cipher cipher = createEncryptCipher();
+        byte[] iv = cipher.getIV();
+        ByteBuffer output = ByteBuffer
+                .allocate(iv.length + cipher.getOutputSize(data.remaining()))
+                .put(iv);
+        try {
+            cipher.doFinal(data, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to encrypt data: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public byte[] encrypt(byte[] data)
+    {
+        return unwrapOrCopyToByteArray(encrypt(ByteBuffer.wrap(data)));
+    }
+
+    /**
+     * Setting the {@link AesSpillCipher#key} to null allows the key to be reclaimed by GC at the time of
+     * destruction, even if the {@link AesSpillCipher} itself is still reachable for longer. When the key
+     * is null, subsequent encrypt / decrypt operations will fail and can prevent accidental use beyond the
+     * intended lifespan of the {@link AesSpillCipher}.
+     */
+    @Override
+    public void destroy()
+    {
+        this.key = null;
+    }
+
+    @Override
+    public boolean isDestroyed()
+    {
+        return key == null;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this).add("isDestroyed", isDestroyed()).toString();
+    }
+
+    private SecretKey verifyKeyNotDestroyed()
+    {
+        SecretKey key = this.key;
+        if (key == null) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Spill cipher already destroyed");
+        }
+        return key;
```

Comment:
I see the merits of returning the `key` here, but it doesn't guarantee every access to `key` field goes through this method -- the field remains directly accessible, so it's up to callers to ensure the check is in place when accessing the `key` field.
For this reason, I'd make `verifyKeyNotDestroyed` return `void`.

Reply:
Also, I'd also use ordinary `checkState(key!=null, ...)` -- this is not supposed to ever be seen by the user, so we don't need to throw `PrestoException` here.
Same applies in other places in this file. _I'm fine if you leave this as-is_.

Reply:
I'm going to leave it as-is, but explain the rationale:
- I agree that having the `verifyKeyNotDestroyed()` method doesn't prevent code from accesses the key directly, however- all code currently accesses the key via this method. It's not refactor-proof, but it is repeated checking logic either way.
- I chose to use the null check or return pattern here for one basic reason: while this class doesn't aim to be "threadsafe" in the sense of modifications from one thread being immediately visible from other threads, but does attempt to be coherent between the destruction check and usage of the key. If a `destroy()` call occurs concurrently with a usage, it's possible that `checkState(key != null, ...)` would succeed, but a `NullPointerException` would still occur on the subsequent usage of the key. This pattern avoids that by making sure that you'll always either get a `PrestoException` or the key will be non-null at the usage site.